compiler/vm/vmdef.nim

```diff
@@ -121,17 +319,61 @@ type
     slots*: ptr UncheckedArray[TFullReg]
     currentException*: PNode
     currentLineInfo*: TLineInfo
+
+    # XXX: These are only here as a temporary measure until I continue 
+    # to work on callback handling
+    typeCache*: ptr TypeInfoCache
+    mem*: ptr VmMemoryManager
+
   VmCallback* = proc (args: VmArgs) {.closure.}
 
+  SlotStorage* = object ## A fixed size container for storing registers 
+                        ## that also allows for shallow copies
+    slots: ptr UncheckedArray[TFullReg]
+    length: int
+    owner: bool ## Indicates whether an instance of `SlotStorage` is the owner of the backing memory
+
+  VmMemoryManager* = object
+    # XXX: the current implementation of the memory manager is overly simple and only temporary
+    regions*: seq[tuple[typ: PVmType, count: int, start: int, len: int]]
+    freed*: seq[tuple[start: int, len: int]]
+
+  TypeInfoCache* = object ## An append-only cache for storing all `VmType`s
+                          ## and `PType` to `PVmType` mappings
+    lut*: Table[ItemId, PVmType]
+
+    types*: seq[PVmType]
+
+    # XXX: if the order of the tyBool..tyFloat128 would only include all all 
+    # numeric types (plus bool and char), we could use the `numericTypes` array
+    # insted of use multiple arrays
+    #numericTypes*: array[tyBool..tyFloat128, PVmType] ## A lookup table for all numeric types (ints/float)
+    boolType*: PVmType
+    charType*: PVmType
+    stringType*: PVmType
+    pointerType*: PVmType
+    nodeType*: PVmType
+    intTypes*: array[tyInt..tyInt64, PVmType]
+    uintTypes*: array[tyUInt..tyUInt64, PVmType]
+    floatTypes*: array[tyFloat..tyFloat64, PVmType]
+
+
   PCtx* = ref TCtx
   TCtx* = object of TPassContext # code gen context
     code*: seq[TInstr]
     debug*: seq[TLineInfo]  # line info for every instruction; kept separate
                             # to not slow down interpretation
-    globals*: PNode         #
-    constants*: PNode       # constant data
-    types*: seq[PType]      # some instructions reference types (e.g. 'except')
-    currentExceptionA*, currentExceptionB*: PNode
+    sframes*: seq[TStackFrame] ## The stack of the currently running code # XXX: rename to `stack`?
+    globals*: TGlobals      ## Holds vm objects created by vmgen or those create by opcNew during execution.
+                            ## Acts as the "heap" of the VM
+    # XXX: constants need to have fixed location in memory. `ref` is used to accomplish that until I 
+    # decide on a more efficient memory management strategy
+    constants*: seq[ref TypedVmValue]       # constant data
+    typeInfoCache*: TypeInfoCache ## Manages the types used the VM
+    types*: seq[tuple[nType: PType, layoutDesc: PVmType]]      # some instructions reference types (e.g. 'except')
```

Comment:
What does this field do, exactly? Isn't `typeInfoCache` supposed to be your one-stop-shop for all types?

Reply:
The `typeInfoCache` manages mappings between `PType` -> `PVmType` and also stores all created `PVmType`s. The `types` field is a lookup list for instructions that require type information (e.g. `opcNew`)

Reply:
Uh... how is it any different from typeInfoCache if it's also a lookup list?

Reply:
`types` provides `range[0..regBxMax] -> (PType, PVmType)` mappings while `typesInfoCache` provides `PType -> PVmType` mappings.

Maybe it's easier if you describe how you think it should be done and I then tell you if it would work or, if not, why it wouldn't?

Reply:
Oh, so that's why. I was wondering why you have two storage for basically the same stuff.

I guess we can use only one storage once you figured the way to get rid of `PVmType` in favor of a central index-based storage.

Reply:
Should I do the `ref`-based to index-based type management change as part of this PR? It's not too hard, but would require another larger scale refactoring.